Add support for pubsub messages

zmacks opened the pull request:
This PR does three things:
1. Creates a new API, "/pubsubLogs"
2. Adds a new function called "trackLogRequest"
3. Refactors "track" to call "trackLogRequest"

Descriptions:
- /pubsubLogs extracts base64 encoded contents from a message envelope, decodes the message, and calls "trackLogRequest"
- "trackLogRequest" takes a LogRequest object instead of a Context object and handles all the logic that "track" used to.
- "track" now calls "trackLogRequest" after validating that "single" and "multiple" are not both null.

Current status: Done. Tests pass locally. Needs some linting work.

Commit:
Initial attempt, tests dont pass yet

Commit:
Small tweaks

zmacks committed:
Decode messages, in a running state but needs more attention

zmacks committed:
Create trackLogRequest endpoint

zmacks committed:
Remove comments

zmacks committed:
Satisfy linter

zmacks committed:
More linter pleasing

zmacks committed:
Linter stuff

zmacks committed:
Linter stuff

zmacks committed:
Linter stuff

zmacks committed:
Linter stuff

zmacks committed:
Linter stuff

naddeoa commented on build.gradle.kts:
```diff
@@ -93,6 +93,8 @@ dependencies {
     implementation("com.fasterxml.jackson.core:jackson-databind:2.10.1")
 
+    implementation("com.google.code.gson:gson:2.9.1")
+
```
It looks good to me overall but is there a reason we need Gson? We should already have jackson in there.

Commit:
PR feedback for pubsub messages

- Added more error handling around the base64+json decoding
- Updated examples to have a valid base64 LogRequest
- Centralized request validation in the new trackRequest
- Changed some variable names/casing

Commit:
Added smoke test for pubsub

This will make sure it actually properly handles/decodes real messages.

naddeoa reviewed (commented):
I added a new commit with a bunch of the stuff I was going to say. I figured you probably don't work in Kotlin all the time so I'll just propose some concrete changes. LMK if looks ok. I called out what I did in the commits.

zmacks committed:
Add publishTime and orderingKey fields to Message data class

zmacks committed:
Fix lint issue

zmacks commented on src/main/kotlin/ai/whylabs/services/whylogs/core/WhyLogsController.kt:
```diff
@@ -281,8 +281,18 @@ data class PubSubEnvelope(
     val attributes: Map<String, String>? = null,
-    @Schema(example = "ewogICAgImRhdGFzZXRJZCI6ICJkZW1vLW1vZGVsIiwKICAgICJ0aW1lc3RhbXAiOiAxNjQ4MTYyNDk0OTQ3LAogICAgInRhZ3MiOiB7CiAgICAgICAgInRhZzEiOiAidmFsdWUxIgogICAgfSwKICAgICJtdWx0aXBsZSI6IHsKICAgICAgICAiY29sdW1ucyI6IFsKICAgICAgICAgICAgIkJyYW5kIiwKICAgICAgICAgICAgIlByaWNlIgogICAgICAgIF0sCiAgICAgICAgImRhdGEiOiBbCiAgICAgICAgICAgIFsgIkhvbmRhIENpdmljIiwgMjIwMDAgXSwKICAgICAgICAgICAgWyAiVG95b3RhIENv ...
+    @Schema(
+        description = "The message data field. If this field is empty, the message must contain at least one attribute.",
+        example = "ewogICAgImRhdGFzZXRJZCI6ICJkZW1vLW1vZGVsIiwKICAgICJ0aW1lc3RhbXAiOiAxNjQ4MTYyNDk0OTQ3LAogICAgInRhZ3MiOiB7CiAgICAgICAgInRhZzEiOiAidmFsdWUxIgogICAgfSwKICAgICJtdWx0aXBsZSI6IHsKICAgICAgICAiY29sdW1ucyI6IFsKICAgICAgICAgICAgIkJyYW5kIiwKICAgICAgICAgICAgIlByaWNlIgogICAgICAgIF0sCiAgICAgICAgImRhdGEiOiBbCiAgICAgICAgICAgIFsgIkhvbmRhIENpdmljIiwgMjIwMDAgXSwKICAgICAgICAgICAgWyAiVG95b3RhIENvcm9s ...
+        )
     val data: String,
     @Schema(example = "123")
-    val messageId: String
+    val messageId: String,
+    @Schema(
+        description = "A timestamp in RFC3339 UTC 'Zulu' format, with nanosecond resolution and up to nine fractional digits",
+        example = "2014-10-02T15:01:23Z"
+        )
+    val publishTime: String,
+    @Schema(description = "If non-empty, identifies related messages for which publish order should be respected")
+    val orderingKey: String?
```
I left out two fields of the Message schema, so I added them in this commit, FYI @naddeoa

naddeoa replied:
Do you think we'll end up needing to use these fields inside of the container at some point? I added a config that should allow extra fields that we don't know about, in case you were just worried about serialization errors @zmacks

zmacks replied:
I don't think we will need to use them, necessarily. At least not right now. Yeah, I was worried about serialization errors so that's a nice peace of mind. Thanks @naddeoa. I think everything looks good!

zmacks committed:
Update test

zmacks committed:
Remove trailing spaces

Commit:
Fix tests + some bugs in the new pubsub api

Forgot to add the ::after call for the main api, which meant that
requests were never consumed when using in-memory queueing

naddeoa reviewed (approved):
Look good. Thanks for being the first contributor!
